### PR TITLE
Fixed #24899 -- Split migrations autodetector._detect_changes() method

### PR DESCRIPTION
**Pull request by iamsteadman**
This would be my first functional commit. I've kept the underscore notation as these are methods that run as part of `_detect_changes`, and I've moved the comments into the individual methods.

### CODE REVIEW

File: django/db/migrations/autodetector.py [hunk 1]
```diff
@@ -220,15 +193,74 @@ def _detect_changes(self, convert_apps=None, graph=None):
...
+
+        return self.migrations
+
+    def _prepare_field_lists(self):
+        """
+        Prepare field lists, and prepare a list of the fields that used
+        through models in the old state so we can make dependencies
+        from the through model deletion to the field that uses it.
+        """
+
```
Comment (timgraham) [inline django/db/migrations/autodetector.py]: I'd omit a blank line after each docstring.

**Comment by timgraham** [comment]
Please follow the commit message guidelines as noted in our [patch review checklist](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist).

**Comment by timgraham** [comment]
Want to squashing commits as described here: https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/working-with-git/#rebasing-branches ?

File: django/db/migrations/autodetector.py [hunk 2]
```diff
@@ -207,7 +181,6 @@ def _detect_changes(self, convert_apps=None, graph=None):
         # isn't bad, but we need to pull a few things around so FKs work nicely
         # inside the same app
         for app_label, ops in sorted(self.generated_operations.items()):
-
             # construct a dependency graph for intra-app dependencies
```
Comment (MarkusH) [inline django/db/migrations/autodetector.py]: I'd move this block into e.g. `_sort_operations()`

File: django/db/migrations/autodetector.py [hunk 3]
```diff
@@ -220,16 +193,72 @@ def _detect_changes(self, convert_apps=None, graph=None):
...
-        # We do this by stepping up an app's list of operations until we
-        # find one that has an outgoing dependency that isn't in another app's
-        # migration yet (hasn't been chopped off its list). We then chop off the
-        # operations before it into a migration and move onto the next app.
-        # If we loop back around without doing anything, there's a circular
-        # dependency (which _should_ be impossible as the operations are all
-        # split at this point so they can't depend and be depended on)
+        self._build_migration_list(graph)
+
+        # OK, add in internal dependencies among the migrations
```
Comment (MarkusH) [inline django/db/migrations/autodetector.py]: I'd move this and the remaining 2 blocks into `_optimize_migrations()`.

**Comment by iamsteadman** [comment]
Hope this is a little better @MarkusH

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by timgraham** [comment]
merged in ed63df3fc2f0c228354ca3435e9bed45be9d0909, thanks!

**Comment by iamsteadman** [comment]
Awesome! Thanks for your patience @timgraham

**Pull request closed without merging**